### DISCUSSION TITLE
Is the semantic of `into sqlite` inconsistent?

### DISCUSSION THREAD

**Discussion by Nicoretti** [nushell/nushell#11175] [General]
Hi @ all,

I've recently started exploring `nu-shell` and came across this [PR](https://github.com/nushell/nushell/pull/10232) while doing some research. I really like that change, however, I found it a bit counterintuitive that using `into sqlite` doesn't generate any output despite creating a database file. When compared to other `into` subcommands, this behavior seems quite different.

Wouldn't it make more sense for it to, for instance, generate the `CREATE TABLE` statement and/or `INSERT INTO` statements required to create a SQLite database?

Personally, I would expect a command like `save` to handle the creation of a database file, such as `ls | save db my.db`.

I'd love to hear different perspectives and explanations regarding this.

Best
Nico

**Comment by fdncred** [comment]
Long ago we had some sqlite commands, when I wrote `into sqlite` I wrote it with one of them in mind. 

I agree with you that `into sqlite` is an outlier and behaves differently than other `into` commands but it still seemed the most intuitive to me for reading. `ls | into sqlite file.db` reads take the output of `ls` and put it `into` a `sqlite` database with file name `file.db`.

I'm not sure `save db` reads the same way since `db` could be anything. Although `save sqlite` would probably work, just renaming the command. 

I still think the sqlite functionality needs to be enhanced in nushell. This command is pretty hacky and would love to see more robust sqlite building.

**Comment by Nicoretti** [comment]
Hi @fdncred,

Thanks for the timely and insightful response.

I definitely agree that from a reading point of view, `ls | into sqlite file.db` is good. However, I believe that inconsistencies make it harder to find and deal with for newcomers like me 😅. I think a congruent interface also helps to ease self-discoverability. Therefore, I would be a huge fan of `save sqlite ...`.

That said, I like a lot of the concepts & ideas of the `nushell` and will definitely take some time to get deeper into it.
Thanks to you and all the other contributors for providing this nice piece of software! 👍

**Reply by fdncred** [reply]
Thanks. I'm currently wondering if `save sqlite` is not appropriate either because it wouldn't play well with the other sqlite commands like `schema`, `query db`, and I just added a PR for a `stor` family of commands for working with in-memory sqlite databases. I'll have to think on it some more.

**Reply by Nicoretti** [reply]
In the end you have more experience with all the commands, so you understand them and the overall concepts and structure better. So I am convinced you gonna take a good decision.

I just can tell you, what confuses me s a beginner: 

In constructing pipelines, I personally see three main "concepts/players": `source`, `filter`, and `sink`. I expected these categories to align with the command names: `from` (source), `to` (filter), and `save` (sink). However, this alignment only seems partially valid, especially when considering the `filter` command.

- `into`: Commands used to convert data from one type to another.
- `to`: Translates structured data to a specific format.

Both commands share similar semantics, differing only in the sense that one deals with format conversion while the other translates types. This realization makes me wonder if there are other command groups for different types of conversions. If so, consulting a specific help/manual page might not lead me to the conversion I'm seeking.

I see a similarity issue regarding `save` and `store`. They seem to address the same abstract concept.
Hard to tell whats different between a `save` and `store` without any additional information.

I am wondering if I missed something regarding what connects the dots or the reasoning behind having those similar yet different commands.

Best,
Nico

**Reply by fdncred** [reply]
I can understand your confusion in all this. We're making it up as we go and have a bunch of contributors. Not everyone thinks the same way, so we usually put things up in a PR and adjust later. Most of the time it's easier to write the code than to name the command you're working on. So, I'm not super worried about having the name right when I first write a command. It's easy enough to change it later.

`stor` is just a way to say store information in an in-memory sqlite database. I needed a command name to based subcommands off of and that's what I came up with. It's not `save`. I always think of `save` as writing to a file. I also wanted to differentiate it with our ability to open and explore sqlite database files. We have `open`, `schema`, and `query db` that can do that. We used to have a bunch of commands to formulate sql calls but we removed them because it was too confusing to have separate `select`, `where`, and team.

I agree with you that we need to get our story right with the naming convention we use for this stuff. I'm just not sure what "right" looks like yet.

**Reply by Nicoretti** [reply]
I can see your points. Thanks for the detailed insights and feedback, highly appreciate it! :+1:

~~Even though you likely already thought about it and decided against it, `commit` could be a possible alternative for `stor` (just a thought).~~ (just saw the PR, does not make sense in context of the other commands)

I will leave it at that and will keep tinkering on my first Nu plugin :wink: .
